Update ROSE to v1.1.3

kylegodbey opened the pull request:
This represents the changes ROSE has seen over the last year of development. Full changelogs and info are listed in the release page. We have a few active actions that build the package, docs, as well as some analytics collectors. Tests can be ran manually, though they are also triggered by the github actions, so you can check the run history for a log.

wildsm commented on CHANGELOG.rst:
```diff
@@ -15,6 +15,7 @@ New capabilities and notable changes:
 
 - added BAND-compatible PUQ, a parallel package for generating experimental designs tailored for uncertainty quantification at `v0.1.0 <https://github.com/parallelUQ/PUQ/releases/tag/v0.1.0>`_
 - updated BAND-compatible parMOO to `v0.4.1 <https://github.com/parmoo/parmoo/releases/tag/v0.4.1>`_, which now includes JIT compilation and automatic differentiation capabilities via `jax`
+- updated BAND-compatible rose to `v1.1.3 <https://github.com/bandframework/rose/releases/tag/v1.1.3>`_
```
@kylegodbey - do you want to add a very short snippet to highlight key change(s) (see parmoo as example)?

kylegodbey replied:
I added a snippet from our v1.1.0 release notes on some highlights

wildsm reviewed (commented):
The rosebandsdk.md  under `M2` mentions 

```
A basic example is explicitly shown in the README
```
but I do not see this. I just see a link to the generic tutorials directory (https://github.com/bandframework/rose/tree/cd96112920d4caf3aefedcdfd47f5b75c5f2554a/docs/tutorials)

kylegodbey commented:
Let me tag @beykyle as well so he can follow along and tackle any of the review notes.

mosesyhc reviewed (commented):
I tested rose by the following:

- Successfully installed rose in developer mode.
- Ran the three tutorial jupyter notebooks. 

Minor comments/questions, that do not impact ROSE's functionality: 
- For the tutorial, `usetex=True` will throw an error for machines without a TeX installation.  
- In an unlikely situation, what if some user is interested in using `nuclear-rose` and `rose` from PyPI at the same time?  Both will take up `import rose` in initialization.

kylegodbey commented:
> I tested rose by the following:
> 
> * Successfully installed rose in developer mode.
> * Ran the three tutorial jupyter notebooks.
> 
> Minor comments/questions, that do not impact ROSE's functionality:
> 
> * For the tutorial, `usetex=True` will throw an error for machines without a TeX installation.
> * In an unlikely situation, what if some user is interested in using `nuclear-rose` and `rose` from PyPI at the same time?  Both will take up `import rose` in initialization.

Regarding TeX, are you aware of any alternatives or checks that could be done to not have TeX as a dependency? It can be quite large for a full installation so we should probably avoid that. I'll do some tests in a clean codespaces as well.

Regarding the import, that package was last touched 11 years ago. They only mention Python 2.5+ as compatible, so not sure about Python 3 compatibility. If it's not, then I don't think we'd run into that conflict just based on our required version.

mosesyhc commented:
> > I tested rose by the following:
> > 
> > * Successfully installed rose in developer mode.
> > * Ran the three tutorial jupyter notebooks.
> > 
> > Minor comments/questions, that do not impact ROSE's functionality:
> > 
> > * For the tutorial, `usetex=True` will throw an error for machines without a TeX installation.
> > * In an unlikely situation, what if some user is interested in using `nuclear-rose` and `rose` from PyPI at the same time?  Both will take up `import rose` in initialization.
> 
> Regarding TeX, are you aware of any alternatives or checks that could be done to not have TeX as a dependency? It can be quite large for a full installation so we should probably avoid that. I'll do some tests in a clean codespaces as well.
> 
> Regarding the import, that package was last touched 11 years ago. They only mention Python 2.5+ as compatible, so not sure about Python 3 compatibility. If it's not, then I don't think we'd run into that conflict just based on our required version.

Simply replace `usetex=True` with `usetex=False` everywhere we set up the plots.  We will have to remove the serif fonts specifications to avoid warnings.

Makes sense to me.  I didn't think that would be too big of an issue.

kylegodbey commented:
> Note that when I follow the developer instructions in the README I get:
> 
> ```
> $ pip install -e .
> Defaulting to user installation because normal site-packages is not writeable
> Obtaining file:///home/wild/oldcomp/repos/bandframework/software/rose
>   Installing build dependencies ... done
>   Checking if build backend supports build_editable ... done
> ERROR: Project file:///home/wild/oldcomp/repos/bandframework/software/rose has a 'pyproject.toml' and its build backend is missing the 'build_editable' hook. Since it does not have a 'setup.py' nor a 'setup.cfg', it cannot be installed in editable mode. Consider using a build backend that supports PEP 660.
> ```

@wildsm when I build from the rose repo it works -- is there any reason why a running from the submodule would break a build?

mosesyhc commented:
> > Note that when I follow the developer instructions in the README I get:
> > ```
> > $ pip install -e .
> > Defaulting to user installation because normal site-packages is not writeable
> > Obtaining file:///home/wild/oldcomp/repos/bandframework/software/rose
> >   Installing build dependencies ... done
> >   Checking if build backend supports build_editable ... done
> > ERROR: Project file:///home/wild/oldcomp/repos/bandframework/software/rose has a 'pyproject.toml' and its build backend is missing the 'build_editable' hook. Since it does not have a 'setup.py' nor a 'setup.cfg', it cannot be installed in editable mode. Consider using a build backend that supports PEP 660.
> > ```
> 
> @wildsm when I build from the rose repo it works -- is there any reason why a running from the submodule would break a build?

The build-backend error may have to do with the `setuptools` versions `<62.0.0`. ? 

Refer https://stackoverflow.com/questions/69711606/how-to-install-a-package-using-pip-in-editable-mode-with-pyproject-toml

wildsm commented:
As a follow up, @mosesyhc is correct: if I locally change `pyproject.toml` to have  `"setuptools>=62.0"`, then pip install succeeds on my testing machine.

Although pytest wasn't mentioned, I figured it might be helpful to share the below results.

```
 pytest
============================= test session starts ==============================
platform linux -- Python 3.10.12, pytest-8.1.1, pluggy-1.4.0
rootdir: /home/wild/oldcomp/repos/bandframework/software/rose
configfile: pyproject.toml
plugins: cov-5.0.0, dash-2.17.1, anyio-4.3.0
collected 6 items                                                              

tests/test_interactions.py .                                             [ 16%]
tests/test_kd.py FF                                                      [ 50%]
tests/test_sae.py F                                                      [ 66%]
tests/test_schroedinger_equation.py .F                                   [100%]

=================================== FAILURES ===================================
_____________________ TestKDNeutron.test_kd_params_default _____________________

self = <test_kd.TestKDNeutron testMethod=test_kd_params_default>

    def test_kd_params_default(self):
>       potential = rose.koning_delaroche.KDGlobal(rose.Projectile.neutron)
E       AttributeError: module 'rose.koning_delaroche' has no attribute 'KDGlobal'

tests/test_kd.py:16: AttributeError
_____________________ TestKDProton.test_kd_params_default ______________________

self = <test_kd.TestKDProton testMethod=test_kd_params_default>

    def test_kd_params_default(self):
>       potential = rose.koning_delaroche.KDGlobal(rose.Projectile.proton)
E       AttributeError: module 'rose.koning_delaroche' has no attribute 'KDGlobal'

tests/test_kd.py:62: AttributeError
____________________________ TestSAE.test_emulation ____________________________

self = <test_sae.TestSAE testMethod=test_emulation>

    def test_emulation(self):
        AMU = 931.5
        A = 40
        mu = A / (A + 1) * AMU
        energy = 13.659
    
        alphastar = np.array(
            [
                51.9,
                4.00134372,
                0.75,
                0.4,
                4.5143365,
                0.51,
                8.3,
                4.5143365,
                0.51,
                6.2,
                3.45415141,
                0.75,
                1.0,
                3.45415141,
                0.75,
            ]
        )
        ntheta = alphastar.size
    
        train = np.load(test_dir / "train.npy")
    
>       interaction = rose.InteractionEIMSpace(
            coordinate_space_potential=rose.koning_delaroche.KD_simple,
            spin_orbit_term=rose.koning_delaroche.KD_simple_so,
            n_theta=ntheta,
            mu=mu,
            energy=energy,
            is_complex=True,
            training_info=train,
            n_basis=15,
            explicit_training=True,
        )
E       TypeError: InteractionEIMSpace.__init__() got an unexpected keyword argument 'spin_orbit_term'

tests/test_sae.py:47: TypeError
____________________________ TestSchrEq.test_energy ____________________________

self = <test_schroedinger_equation.TestSchrEq testMethod=test_energy>

    def test_energy(self):
        """
        Tests the correct handling of energy-dependent interactions. Does not
        use emulation.
        Compares the HF solutions from two interactions -- one that is
        energy-dependent and one that is not.
        """
        mu = 1.0
        energy = 50.0
        ell = 0
        z = 0
        rho = rose.constants.DEFAULT_RHO_MESH
    
        @njit
        def potential(r, theta):
            return theta[0] * rose.utility.woods_saxon_safe(r, theta[1], theta[2])
    
>       interaction = rose.Interaction(
            coordinate_space_potential=potential,
            n_theta=3,
            mu=mu,
            energy=energy,
            Z_1=z,
            Z_2=z,
        )
E       TypeError: Interaction.__init__() missing 1 required positional argument: 'ell'

tests/test_schroedinger_equation.py:46: TypeError
=========================== short test summary info ============================
FAILED tests/test_kd.py::TestKDNeutron::test_kd_params_default - AttributeError: module 'rose.koning_delaroche' has no attribute 'KDGlobal'
FAILED tests/test_kd.py::TestKDProton::test_kd_params_default - AttributeError: module 'rose.koning_delaroche' has no attribute 'KDGlobal'
FAILED tests/test_sae.py::TestSAE::test_emulation - TypeError: InteractionEIMSpace.__init__() got an unexpected keyword argumen...
FAILED tests/test_schroedinger_equation.py::TestSchrEq::test_energy - TypeError: Interaction.__init__() missing 1 required positional argument: '...
========================= 4 failed, 2 passed in 3.16s ==========================
```

wildsm reviewed (commented):
As a follow up, @mosesyhc is correct: if I locally change `pyproject.toml` to have  `"setuptools>=62.0"`, then pip install succeeds on my testing machine.

Although pytest wasn't mentioned, I figured it might be helpful to share the below results.

```
 pytest
============================= test session starts ==============================
platform linux -- Python 3.10.12, pytest-8.1.1, pluggy-1.4.0
rootdir: /home/wild/oldcomp/repos/bandframework/software/rose
configfile: pyproject.toml
plugins: cov-5.0.0, dash-2.17.1, anyio-4.3.0
collected 6 items                                                              

tests/test_interactions.py .                                             [ 16%]
tests/test_kd.py FF                                                      [ 50%]
tests/test_sae.py F                                                      [ 66%]
tests/test_schroedinger_equation.py .F                                   [100%]

=================================== FAILURES ===================================
_____________________ TestKDNeutron.test_kd_params_default _____________________

self = <test_kd.TestKDNeutron testMethod=test_kd_params_default>

    def test_kd_params_default(self):
>       potential = rose.koning_delaroche.KDGlobal(rose.Projectile.neutron)
E       AttributeError: module 'rose.koning_delaroche' has no attribute 'KDGlobal'

tests/test_kd.py:16: AttributeError
_____________________ TestKDProton.test_kd_params_default ______________________

self = <test_kd.TestKDProton testMethod=test_kd_params_default>

    def test_kd_params_default(self):
>       potential = rose.koning_delaroche.KDGlobal(rose.Projectile.proton)
E       AttributeError: module 'rose.koning_delaroche' has no attribute 'KDGlobal'

tests/test_kd.py:62: AttributeError
____________________________ TestSAE.test_emulation ____________________________

self = <test_sae.TestSAE testMethod=test_emulation>

    def test_emulation(self):
        AMU = 931.5
        A = 40
        mu = A / (A + 1) * AMU
        energy = 13.659
    
        alphastar = np.array(
            [
                51.9,
                4.00134372,
                0.75,
                0.4,
                4.5143365,
                0.51,
                8.3,
                4.5143365,
                0.51,
                6.2,
                3.45415141,
                0.75,
                1.0,
                3.45415141,
                0.75,
            ]
        )
        ntheta = alphastar.size
    
        train = np.load(test_dir / "train.npy")
    
>       interaction = rose.InteractionEIMSpace(
            coordinate_space_potential=rose.koning_delaroche.KD_simple,
            spin_orbit_term=rose.koning_delaroche.KD_simple_so,
            n_theta=ntheta,
            mu=mu,
            energy=energy,
            is_complex=True,
            training_info=train,
            n_basis=15,
            explicit_training=True,
        )
E       TypeError: InteractionEIMSpace.__init__() got an unexpected keyword argument 'spin_orbit_term'

tests/test_sae.py:47: TypeError
____________________________ TestSchrEq.test_energy ____________________________

self = <test_schroedinger_equation.TestSchrEq testMethod=test_energy>

    def test_energy(self):
        """
        Tests the correct handling of energy-dependent interactions. Does not
        use emulation.
        Compares the HF solutions from two interactions -- one that is
        energy-dependent and one that is not.
        """
        mu = 1.0
        energy = 50.0
        ell = 0
        z = 0
        rho = rose.constants.DEFAULT_RHO_MESH
    
        @njit
        def potential(r, theta):
            return theta[0] * rose.utility.woods_saxon_safe(r, theta[1], theta[2])
    
>       interaction = rose.Interaction(
            coordinate_space_potential=potential,
            n_theta=3,
            mu=mu,
            energy=energy,
            Z_1=z,
            Z_2=z,
        )
E       TypeError: Interaction.__init__() missing 1 required positional argument: 'ell'

tests/test_schroedinger_equation.py:46: TypeError
=========================== short test summary info ============================
FAILED tests/test_kd.py::TestKDNeutron::test_kd_params_default - AttributeError: module 'rose.koning_delaroche' has no attribute 'KDGlobal'
FAILED tests/test_kd.py::TestKDProton::test_kd_params_default - AttributeError: module 'rose.koning_delaroche' has no attribute 'KDGlobal'
FAILED tests/test_sae.py::TestSAE::test_emulation - TypeError: InteractionEIMSpace.__init__() got an unexpected keyword argumen...
FAILED tests/test_schroedinger_equation.py::TestSchrEq::test_energy - TypeError: Interaction.__init__() missing 1 required positional argument: '...
========================= 4 failed, 2 passed in 3.16s ==========================
```

mosesyhc commented:
hmm... for what's worth, I also attempted `pytest`.  They passed, with warnings.  See the output below.

```
$ pytest
=============================================================================================================== test session starts ===============================================================================================================
platform win32 -- Python 3.11.9, pytest-8.3.3, pluggy-1.5.0
rootdir: C:\Users\cmyh\Desktop\git\rose
configfile: pyproject.toml
collected 6 items                                                                                                                                                                                                                                  

tests\test_interactions.py .                                                                                                                                                                                                                 [ 16%] 
tests\test_kd.py ..                                                                                                                                                                                                                          [ 50%]
tests\test_sae.py .                                                                                                                                                                                                                          [ 66%]
tests\test_schroedinger_equation.py ..                                                                                                                                                                                                       [100%]

================================================================================================================ warnings summary ================================================================================================================= 
..\bandframework\software\rose\src\rose\utility.py:322
  C:\Users\cmyh\Desktop\git\bandframework\software\rose\src\rose\utility.py:322: DeprecationWarning: invalid escape sequence '\s'
    __AME_DB__ = pd.read_csv(__AME_PATH__, sep="\s+")

..\bandframework\software\rose\src\rose\training.py:391
  C:\Users\cmyh\Desktop\git\bandframework\software\rose\src\rose\training.py:391: DeprecationWarning: invalid escape sequence '\p'
    def multiple_formatter(denominator=2, number=np.pi, latex="\pi"):

..\bandframework\software\rose\src\rose\training.py:423
  C:\Users\cmyh\Desktop\git\bandframework\software\rose\src\rose\training.py:423: DeprecationWarning: invalid escape sequence '\p'
    def __init__(self, denominator=2, number=np.pi, latex="\pi"):

tests/test_sae.py: 62 warnings
tests/test_schroedinger_equation.py: 2 warnings
  C:\Users\cmyh\Desktop\git\bandframework\software\rose\src\rose\free_solutions.py:53: DeprecationWarning: scipy.misc.derivative is deprecated in SciPy v1.10.0; and will be completely removed in SciPy v1.12.0. You may consider using findiff: https://github.com/maroba/findiff or numdifftools: https://github.com/pbrod/numdifftools
    return derivative(lambda z: H_minus(z, ell, eta), rho, dx=dx)

tests/test_sae.py: 62 warnings
tests/test_schroedinger_equation.py: 2 warnings
  C:\Users\cmyh\Desktop\git\bandframework\software\rose\src\rose\free_solutions.py:46: DeprecationWarning: scipy.misc.derivative is deprecated in SciPy v1.10.0; and will be completely removed in SciPy v1.12.0. You may consider using findiff: https://github.com/maroba/findiff or numdifftools: https://github.com/pbrod/numdifftools
    return derivative(lambda z: H_plus(z, ell, eta), rho, dx=dx)

tests/test_sae.py: 15 warnings
  C:\Users\cmyh\Desktop\git\bandframework\software\rose\src\rose\utility.py:305: DeprecationWarning: scipy.misc.derivative is deprecated in SciPy v1.10.0; and will be completely removed in SciPy v1.12.0. You may consider using findiff: https://github.com/maroba/findiff or numdifftools: https://github.com/pbrod/numdifftools
    return -((1 - x**2) ** (1 / 2)) * derivative(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=================================================================================================== 6 passed, 146 warnings in 133.90s (0:02:13) =================================================================================================== 
```

kylegodbey commented:
Alright, I believe I've addressed the concerns from @mosesyhc (and huge thanks for the PRs!), as well as @wildsm. The only outstanding concern is the tests failing on his machine. The tests work on codespaces, Moses's machine, and the automated test system so I'm really not sure what could cause it other than some files being out of date. Might be worth a wipe and re-pull the submodule.

mosesyhc commented:
@kylegodbey : On my end it's good to go.  I will put in my approval.  It would be nice if refreshing the submodule resolves the issue for @wildsm .

wildsm reviewed (approved):
Correct, the latest version succeeds with `pip install -e .` and running `pytest`  in `tests/` produces only warnings (the matplotlib one can be ignored):

```
============================= test session starts ==============================
platform linux -- Python 3.10.12, pytest-8.1.1, pluggy-1.4.0
rootdir: /home/wild/oldcomp/repos/bandframework/software/rose
configfile: pyproject.toml
plugins: cov-5.0.0, dash-2.17.1, anyio-4.3.0
collected 6 items                                                              

test_interactions.py .                                                   [ 16%]
test_kd.py ..                                                            [ 50%]
test_sae.py 
.                                                            [ 66%]
test_schroedinger_equation.py ..                                         [100%]

=============================== warnings summary ===============================
../src/rose/utility.py:322
  /home/wild/oldcomp/repos/bandframework/software/rose/src/rose/utility.py:322: DeprecationWarning: invalid escape sequence '\s'
    __AME_DB__ = pd.read_csv(__AME_PATH__, sep="\s+")

../src/rose/training.py:391
  /home/wild/oldcomp/repos/bandframework/software/rose/src/rose/training.py:391: DeprecationWarning: invalid escape sequence '\p'
    def multiple_formatter(denominator=2, number=np.pi, latex="\pi"):

../src/rose/training.py:423
  /home/wild/oldcomp/repos/bandframework/software/rose/src/rose/training.py:423: DeprecationWarning: invalid escape sequence '\p'
    def __init__(self, denominator=2, number=np.pi, latex="\pi"):

../../../../../../.local/lib/python3.10/site-packages/matplotlib/projections/__init__.py:63
  /home/wild/.local/lib/python3.10/site-packages/matplotlib/projections/__init__.py:63: UserWarning: Unable to import Axes3D. This may be due to multiple versions of Matplotlib being installed (e.g. as a system package and as a pip package). As a result, the 3D projection is not available.
    warnings.warn("Unable to import Axes3D. This may be due to multiple versions of "

tests/test_sae.py: 62 warnings
tests/test_schroedinger_equation.py: 2 warnings
  /home/wild/oldcomp/repos/bandframework/software/rose/src/rose/free_solutions.py:53: DeprecationWarning: scipy.misc.derivative is deprecated in SciPy v1.10.0; and will be completely removed in SciPy v1.12.0. You may consider using findiff: https://github.com/maroba/findiff or numdifftools: https://github.com/pbrod/numdifftools
    return derivative(lambda z: H_minus(z, ell, eta), rho, dx=dx)

tests/test_sae.py: 62 warnings
tests/test_schroedinger_equation.py: 2 warnings
  /home/wild/oldcomp/repos/bandframework/software/rose/src/rose/free_solutions.py:46: DeprecationWarning: scipy.misc.derivative is deprecated in SciPy v1.10.0; and will be completely removed in SciPy v1.12.0. You may consider using findiff: https://github.com/maroba/findiff or numdifftools: https://github.com/pbrod/numdifftools
    return derivative(lambda z: H_plus(z, ell, eta), rho, dx=dx)

tests/test_sae.py: 15 warnings
  /home/wild/oldcomp/repos/bandframework/software/rose/src/rose/utility.py:305: DeprecationWarning: scipy.misc.derivative is deprecated in SciPy v1.10.0; and will be completely removed in SciPy v1.12.0. You may consider using findiff: https://github.com/maroba/findiff or numdifftools: https://github.com/pbrod/numdifftools
    return -((1 - x**2) ** (1 / 2)) * derivative(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================= 6 passed, 147 warnings in 178.06s (0:02:58) ==================
```
Merging
